Force port 0 in I2P

vasild opened the pull request:
_This is an alternative to https://github.com/bitcoin/bitcoin/pull/21514, inspired by https://github.com/bitcoin/bitcoin/pull/21514#issuecomment-815049933. They are mutually exclusive. Just one of them should be merged._

Change assumed ports for I2P to 0 (instead of the default 8333) as this is closer to what actually happens underneath with SAM 3.1 (https://github.com/bitcoin/bitcoin/pull/21514#issuecomment-812632520, https://github.com/bitcoin/bitcoin/pull/21514#issuecomment-816564719).

Don't connect to I2P peers with advertised port != 0 (we don't specify a port to our SAM 3.1 proxy and it always connects to port = 0).

Note, this change:
* ~~Keeps I2P addresses with port != 0 in addrman and relays them to others via P2P gossip. There may be non-bitcoin-core-22.0 peers using SAM 3.2 and for them such addresses may be useful.~~ Edit: last commit amends that behavior to change all ports to 0 in addrman. There are no peers using SAM 3.2 yet and we need to be able to connect to existent peers that are already saved in addrman with port 8333.
* Silently refuses to connect to I2P hosts with port != 0. This is ok for automatically chosen peers from addrman. Not so ok for peers provided via `-addnode` or `-connect` - a user who specifies `foo.b32.i2p:1234` (non zero port) may wonder why "nothing is happening".

Fixes #21389

vasild committed:
net: change assumed I2P port to 0

* When accepting an I2P connection, assume the peer has port 0 instead
  of the default 8333 (for mainnet). It is not being sent to us, so we
  must assume something.
* When deriving our own I2P listen CService use port 0 instead of the
  default 8333 (for mainnet). So that we later advertise it to peers
  with port 0.

In the I2P protocol SAM 3.1 and older (we use 3.1) ports are not used,
so they are irrelevant. However in SAM 3.2 and newer ports are used and
from the point of view of SAM 3.2, a peer using SAM 3.1 seems to have
specified port=0.

vasild committed:
net: change I2P seeds' ports to 0

vasild committed:
net: distinguish default port per network

Change `CChainParams::GetDefaultPort()` to return 0 if the network is
I2P.

vasild committed:
net: do not connect to I2P hosts on port!=0

When connecting to an I2P host we don't specify destination port and it
is being forced to 0 by the SAM 3.1 proxy, so if we connect to the same
host on two different ports, that would be actually two connections to
the same service (listening on port 0).

Fixes https://github.com/bitcoin/bitcoin/issues/21389

vasild commented:
TODO: see if something has to be done about this assert:

https://github.com/bitcoin/bitcoin/blob/8462cd56010a8beadf95d4f3cb1357ca9d88493b/src/bench/addrman.cpp#L91

sipa commented:
I think that's just a way to test that the returned address object isn't empty. Should probably replace it with another test.

laanwj commented:
Concept ACK, earlier I was convinced it would be better to get rid of ports completely for I2P, but at the time I thought it didn't exist as a concept there. I was unaware that the concept *does* exists for newer versions of the protocol.

This is more forward compatible with new versions of `i2pd` and SAM, so seems better to me.

sipa commented on src/net.cpp:
```diff
@@ -2019,8 +2019,9 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             // port, causing a DoS attack as nodes around the network attempt
             // to connect to it fruitlessly.
-            if (addr.GetPort() != Params().GetDefaultPort() && nTries < 50)
+            if (!addr.IsI2P() && addr.GetPort() != Params().GetDefaultPort() && nTries < 50) {
```
I don't think this is exactly what we want. As (with SAM 3.1) we're unable to connect to anything except port 0 (from the perspective of later versions), shouldn't this equally bias away from port 0, as we bias away from the default port for other networks?

In other words: maybe this can be done a bit cleaner by having a network-dependent default port, which is 0 for I2P and the normal port for everything else?

vasild replied:
I think the limitation of only being able to connect to ports that equal 0 is purely internal to the I2P code - it depends on the SAM version being used and higher level code shouldn't care about that or be aware of it. So I planted the refuse-to-connect-to-non-zero-ports [check in i2p.cpp](https://github.com/bitcoin/bitcoin/pull/22112/commits/4401bb56cf3822950b221c3fa47fb0b65312d12f#diff-8945a03581b372d674f34dcb6a63e537343e376ebcd9d8c7c3d94d4366cd9b9dR175-R180).

This allows some day to remove that by only modifying `i2p.cpp` - easier, compared to if this knowledge is spilled in other modules.

A network-dependent default port seems like an unnecessary generalization to me because it is only relevant for I2P. If there was at least one other network that used !=8333 default port, then it would be more appealing :)

Off-scope: I think this entire `if` should be removed.

jonatack replied:
It may be good to update the comment preceding this line to explain the exception for I2P.

vasild replied:
Added a comment.

jonatack replied:
From testing and offline discussion with @vasild, it seems a network-dependent `GetDefaultPort()` is indeed needed.

vasild replied:
So, network-dependent default port seems to be the most elegant solution to [resolve addresses without ports](https://github.com/bitcoin/bitcoin/pull/22112#issuecomment-872224803) properly - e.g. `1.2.3.4` should be treated as `1.2.3.4:8333` and `foo.b32.i2p` as `foo.b32.i2p:0`.

Now this line looks better?


```cpp
if (addr.GetPort() != Params().GetDefaultPort(addr.GetNetwork()) && nTries < 50) {
```

vasild commented:
> ... I was convinced it would be better to get rid of ports completely for I2P, but at the time I thought it didn't exist as a concept there. I was unaware that the concept does exists for newer versions of the protocol.

Me too!

vasild commented:
> TODO: see if something has to be done about this assert:
> 
> https://github.com/bitcoin/bitcoin/blob/8462cd56010a8beadf95d4f3cb1357ca9d88493b/src/bench/addrman.cpp#L91

That seems to be ok as it is - addresses being used in the bench are generated, not taken from real-world addrman. The code that generates them takes care to avoid port 0:

https://github.com/bitcoin/bitcoin/blob/a9435e34457e0bfebd22e574fe63428537948aeb/src/bench/addrman.cpp#L35-L37

So it might as well generate I2P addresses with !=0 ports. Currently it generates only IPv6 addresses.

jonatack reviewed (commented):
Almost-ACK 4401bb56cf3822950b221c3fa47fb0b65312d12f code review, tested rebased to master, with the following comments and modulo test coverage

- I2P local address port is now 0 in getnetworkinfo and -netinfo
- outbound connections are no longer made to any of the 13 I2P addresses in my addrman (all with port 8333), which is a bit of collateral annoyance for these nodes
-  all of these 13 I2P addresses are still in the addrman, e.g. they are returned by -addrinfo and getnodeaddresses 0 i2p, confirming the pull description
- inbound I2P connections succeed (5 inbound I2P peers after 3 minutes, 7 after 15 minutes) and in getpeerinfo the addr and addrbind fields both show "b32.i2p:0"

This is a smaller and easier change than #21514 and more forward-compatible.

It might be a good idea to add tests, e.g. for the points mentioned in https://github.com/bitcoin/bitcoin/pull/21514#issuecomment-815049933 (and anything else you find salient).

vasild committed:
test: ensure I2P ports are handled as expected

vasild commented:
`4401bb56cf...a9cacd29a1`: add a test and a comment

> outbound connections are no longer made to any of the 13 I2P addresses in my addrman (all with port 8333), which is a bit of collateral annoyance for these nodes

Yes, that's really unfortunate and annoying. I wonder if it can be avoided...

jonatack commented:
ISTM with this change we would need a way to manually or automatically convert the existing I2P entries in the addrman from port 8333 to 0.

jonatack commented:
> `4401bb56cf...a9cacd29a1`: add a test and a comment

Did you mean to not add the test to the test runner?

vasild commented:
`a9cacd29a1...d6255891da`: add the newly added test to `test_runner.py` (forgotten in the previous push)

> Add the test to the test runner?

No :hear_no_evil:

vasild commented:
`d6255891da...28f98950ee`: rename the functional test to match the naming convention

jonatack commented:
`test_runner.py#check_script_prefixes()` is unhappy about the name of the test file

jonatack commented:
ACK 28f98950ee89dafcb75b131b810f5a933c2977b3 modulo https://github.com/bitcoin/bitcoin/pull/22112#issuecomment-866677980

jonatack commented:
IRC discussion with @vasild and I regarding whether the remedies here and in #21514 are less desirable than leaving the "double connections" issue unresolved: https://www.erisian.com.au/bitcoin-core-dev/log-2021-06-23.html#l-167

vasild committed:
addrman: reset I2P ports to 0 when loading from disk

This is a temporary change to convert I2P addresses that have propagated
with port 8333 to ones with port 0.

It would cause a problem some day if indeed some bitcoin software is
listening on port 8333 only and rejects connections to port 0 and we are
still using SAM 3.1 which only supports port 0. In this case we would
replace 8333 with 0 and try to connect to such nodes.

This commit should be included in 22.0 and be reverted before 23.0 is
released.

vasild commented:
`28f98950ee..0c6db5b87f`: append a new commit that changes the ports of existent I2P addrman entries to 0.

vasild commented:
`0c6db5b87f...b336f12eda`: no need to lock `CAddrMan::cs` in the `ResetI2PPorts()` method, it is already locked by the caller and in `master` the mutex is not recursive anymore.

vasild commented:
[Coverage report](https://people.freebsd.org/~vd/pr22112_b336f12ed_coverage/modified_and_not_covered.html) for modified lines by this PR and not covered by tests. There is just one line which was also not covered before this PR (this PR changed `Params().GetDefaultPort()` to `PORT_SAM31` in that line).

The new method added in the last commit is [fully covered](https://people.freebsd.org/~vd/pr22112_b336f12ed_coverage/src/addrman.cpp.gcov.html#670) (I don't know what that red minus is on line 731, but it looks bogus).

The last commit `addrman: reset I2P ports to 0 when loading from disk` is somewhat optional, feel free to ACK the last but one commit. From the list in: https://github.com/bitcoin/bitcoin/issues/21389#issuecomment-866925116 this full PR implements option 3. Without the last commit it implements option 2.

jonatack commented on src/addrman.cpp:
```diff
@@ -665,3 +666,88 @@ std::vector<bool> CAddrMan::DecodeAsmap(fs::path path)
+                vvTried[bucket][i] = -1;
+                addr_info = addr_info_newport;
+            } else {
+                vvTried[bucket][i] = -1;
```
b336f12 it looks like this line can be pulled out of the conditional
```diff
             // Reposition from (bucket, i) to (bucket_target, i_target) or delete if the target
             // position is occupied.
+            vvTried[bucket][i] = -1;
 
             if (vvTried[bucket_target][i_target] == -1) {
                 vvTried[bucket_target][i_target] = id;
-                vvTried[bucket][i] = -1;
                 addr_info = addr_info_newport;
             } else {
-                vvTried[bucket][i] = -1;
```

vasild replied:
Did that, but after that changed the code to remove the existent entry (if any), so the above is not relevant any more.

jonatack commented on src/addrman.cpp:
```diff
@@ -665,3 +666,88 @@ std::vector<bool> CAddrMan::DecodeAsmap(fs::path path)
+                addr_info = addr_info_newport;
+            } else {
+                vvTried[bucket][i] = -1;
+                SwapRandom(addr_info.nRandomPos, vRandom.size() - 1);
```
b336f12 would asserts like those seen in `CAddrMan::Delete()` be a good idea?
```diff
@@ -675,6 +675,7 @@ void CAddrMan::ResetI2PPorts()
             if (id == -1) {
                 continue;
             }
+            assert(mapInfo.count(id) != 0);
             auto& addr_info = mapInfo[id];
@@ -716,6 +717,7 @@ void CAddrMan::ResetI2PPorts()
             if (id == -1) {
                 continue;
             }
+            assert(mapInfo.count(id) != 0);
             auto& addr_info = mapInfo[id];
@@ -737,13 +739,13 @@ void CAddrMan::ResetI2PPorts()
             if (vvTried[bucket_target][i_target] == -1) {
                 vvTried[bucket_target][i_target] = id;
                 addr_info = addr_info_newport;
             } else {
+                assert(addr_info.nRefCount == 0);
                 SwapRandom(addr_info.nRandomPos, vRandom.size() - 1);
```

vasild replied:
I think asserts are too excessive - if something goes wrong with this code I don't think it is critical enough to stop the node (that's during startup). Instead of the first two asserts I added a check & early quit. The 3rd suggested assert is not relevant now because I changed those lines to remove the existent entry (not the I2P one).

jonatack replied:
FWIW the asserts passed for me in mainnet testing.

jonatack reviewed (commented):
Semi-ACK b336f12eda9ce3aeefc6d573c4068a27beef8def modulo the issues noted below in https://github.com/bitcoin/bitcoin/pull/22112#issuecomment-871687616.

A couple of suggestions and questions below, along with some proposed edits in https://github.com/jonatack/bitcoin/commit/539a28180b06fa99f81fbd01223ac30e7e4276af

Nice work on the new unit test!

jonatack commented:
Tested this rebased to master, then restarted the node after backing up `peers.dat`:

- `Loaded 65986 addresses from peers.dat`
- The number of I2P addresses reported by `-addrinfo` and `getnodeaddresses 0 i2p` dropped from 15 to 5 (all have port 0 as expected)
- `I2P: SAM session created: my address=zsxwyo6qcn3chqzwxnseusqgsnuw3maqnztkiypyfxtya4snkoka.b32.i2p:0`
- No outbound I2P connections (due to the other peers not running this branch), 4 inbound ones after a minute

Restarted again, no change:

- `Loaded 66102 addresses from peers.dat`
- 5 I2P addresses known (port 0 as expected)
- `I2P: SAM session created: my address=zsxwyo6qcn3chqzwxnseusqgsnuw3maqnztkiypyfxtya4snkoka.b32.i2p:0`
- No outbound I2P connections, 4 inbound ones after a few minutes
- Tried `addpeeraddress` to re-add the I2P addresses, it returns `"success": true` but the address isn't added.

We should test this branch with two nodes connecting to each other.

It's unfortunate to lose I2P addresses due to the bucket repositioning.

jonatack commented:
```$ test/lint/lint-circular-dependencies.sh
A new circular dependency in the form of "chainparams -> i2p -> chainparams" appears to have been introduced.
```

vasild commented:
`b336f12eda...e4684144e9`:

* If a collision occurs due to re-bucketing of the I2P address with the changed port to 0, then remove the other entry, not the I2P one. Looks like I2P addresses are more precious than a random one from a 50k addrman filled mostly with IPv4 addresses.

* If `addnode=foo.b32.i2p` is specified (without port) then we would do something like `Lookup(str, GetDefaultPort())` and would end up with `CService` with address `foo.b32.i2p` and port `8333` which is not what we want (we want port `0`) (thanks, @jonatack for discovering this!). So, implement @sipa's [suggestion](https://github.com/bitcoin/bitcoin/pull/22112#discussion_r644216009) of network-dependent default port.

* Took some suggestions.

laanwj commented:
Added 22.0 milestone (@jonatack mentioned that one of the solutions to the [I2P port problem](https://github.com/bitcoin/bitcoin/issues/21389#issuecomment-866925116) needs to make it in, and I think that makes sense. From what I understand this PR is preferred.)

vasild committed:
doc: mention that we enforce port=0 in I2P

Co-authored-by: Jon Atack <jon@atack.com>

vasild commented:
`36cf2ca392...ba6d18861b`: rebase in order to pull `doc/i2p.md` and add a note to it for I2P ports being forced to 0.

jonatack commented:
Re-reviewing and testing.

vasild commented:
`ba6d18861b...14db11782c`: minor rewording in `doc/i2p.md`

jonatack reviewed (commented):
Tested ACK 14db11782c467c2585c552ec1797ad3b3ef70ab3 modulo the questions below. Rebased to master, backed up `peers.dat`, debug built with `-DEBUG_ADDRMAN` and ran addrman/net/i2p units at each commit that changes code, tested by restarting several times with i2p/tor/addrman logging, all 14 I2P addresses preserved, quickly connected to 10-11 I2P peers (roughly half in and half manual out), -netinfo/-addrinfo/getnetworkinfo/getnodeaddresses/getaddednodeinfo return as expected with port 0, running with the following asserts -- still running a node with them

<details><summary>asserts</summary><p>

```diff
diff --git a/src/addrman.cpp b/src/addrman.cpp
index f2d54c0adb..db0e7c2c2c 100644
--- a/src/addrman.cpp
+++ b/src/addrman.cpp
@@ -710,9 +710,7 @@ void CAddrMan::ResetI2PPorts()
                 continue;
             }
             auto it = mapInfo.find(id);
-            if (it == mapInfo.end()) {
-                return;
-            }
+            assert(it != mapInfo.end());
             auto& addr_info = it->second;
             if (!addr_info.IsI2P() || addr_info.GetPort() == I2P_SAM31_PORT) {
                 continue;
@@ -737,6 +735,7 @@ void CAddrMan::ResetI2PPorts()
 
             // Reposition from i to i_target, removing the entry from i_target (if any).
             ClearNew(bucket, i_target);
+            assert(vvNew[bucket][i_target] == -1);
             vvNew[bucket][i_target] = id;
             vvNew[bucket][i] = -1;
             addr_info = addr_info_newport;
@@ -750,9 +749,7 @@ void CAddrMan::ResetI2PPorts()
                 continue;
             }
             auto it = mapInfo.find(id);
-            if (it == mapInfo.end()) {
-                return;
-            }
+            assert(it != mapInfo.end());
             auto& addr_info = it->second;
             if (!addr_info.IsI2P() || addr_info.GetPort() == I2P_SAM31_PORT) {
                 continue;
@@ -790,10 +787,13 @@ void CAddrMan::ResetI2PPorts()
                 vvNew[new_bucket][new_bucket_i] = old_target_id;
                 ++nNew;
             }
+            assert(vvTried[bucket_target][i_target] == -1);
 
             vvTried[bucket_target][i_target] = id;
             vvTried[bucket][i] = -1;
             addr_info = addr_info_newport;
+            assert(addr_info.fInTried == true);
         }
     }
 }
```
</p></details>

This is much improved over my last review+testing at b336f12 as the issues mentioned in https://github.com/bitcoin/bitcoin/pull/22112#issuecomment-871687616 have been resolved.

I'm not 100% sure about the tried/new management and it may be good for an addrman expert to look over the changes. From what I can tell I don't see any issues, but noodling around adding things like, say, `++nTried` at the bottom of the last loop in `CAddrMan::ResetI2PPorts()`, didn't break anything either.

Questions, should any of these be updated?
- the `-port` config help in `src/init.cpp`
- `GetDefaultPort()` in CConnman::ThreadDNSAddressSeed()
- `GetListenPort()`

TODO: test bootstrapping from hardcoded I2P seeds.

jonatack commented on src/addrman.cpp:
```diff
@@ -699,3 +700,100 @@ std::vector<bool> CAddrMan::DecodeAsmap(fs::path path)
+            }
+            auto it = mapInfo.find(id);
+            if (it == mapInfo.end()) {
+                return;
```
(Asserts run fine for me in my testing so far).

vasild replied:
Asserts look [scary](https://github.com/bitcoin/bitcoin/pull/22112#discussion_r662245348) to me. If those have to be asserts then I would deem this too risky for 22.0. Maybe I am pessimistic but I imagine 22.0 nodes not being able to start because I overlooked something in the I2P ports fiddling which is definitely not critical for node operation :bomb: :skull_and_crossbones:

vasild commented:
> ... noodling around adding things like, say, `++nTried` at the bottom of the last loop in `CAddrMan::ResetI2PPorts()`, didn't break anything either.

This should be detected by `CAddrMan::Check()` (only enabled if `DEBUG_ADDRMAN` is defined):

https://github.com/bitcoin/bitcoin/blob/271155984574a5bba9619d8f6da9bb0606d93f8c/src/addrman.cpp#L438

> Questions, should any of these be updated?
> 
>     * the `-port` config help in `src/init.cpp`

Add something like "not relevant for I2P"?
 
>     * `GetDefaultPort()` in CConnman::ThreadDNSAddressSeed()
> 
>     * `GetListenPort()`

I think those are ok.

vasild commented on src/chainparams.h:
```diff
@@ -8,11 +8,13 @@
 
 #include <chainparamsbase.h>
 #include <consensus/params.h>
+#include <netaddress.h>
```
Removed the changes to `chainparams.cpp`.

vasild commented on src/chainparams.h:
```diff
@@ -80,6 +82,8 @@ class CChainParams
     const Consensus::Params& GetConsensus() const { return consensus; }
     const CMessageHeader::MessageStartChars& MessageStart() const { return pchMessageStart; }
     uint16_t GetDefaultPort() const { return nDefaultPort; }
+    uint16_t GetDefaultPort(Network net) const;
```
Done.

vasild commented on src/chainparams.cpp:
```diff
@@ -540,6 +540,20 @@ const CChainParams &Params() {
+    if (a.SetSpecial(addr)) {
+        return GetDefaultPort(a.GetNetwork());
+    }
+    return GetDefaultPort();
```
Done, and moved to `.h`

jonatack commented:
> > ... noodling around adding things like, say, `++nTried` at the bottom of the last loop in `CAddrMan::ResetI2PPorts()`, didn't break anything either.
> 
> This should be detected by `CAddrMan::Check()` (only enabled if `DEBUG_ADDRMAN` is defined):
> 
> https://github.com/bitcoin/bitcoin/blob/271155984574a5bba9619d8f6da9bb0606d93f8c/src/addrman.cpp#L438

Yes, I'll reverify if `CAddrMan::Check_()` is actually running the way I've been configuring.

jonatack commented:
re-ACK 0b0ee030ea6f2b6508e8660dffc40b40c6aa2644 per `git diff 14db117 0b0ee03`


> * the `-port` config help in `src/init.cpp`
> 
> Add something like "not relevant for I2P"?

sgtm

jonatack commented on src/addrman.cpp:
```diff
@@ -699,3 +700,100 @@ std::vector<bool> CAddrMan::DecodeAsmap(fs::path path)
+{
+    for (int bucket = 0; bucket < ADDRMAN_NEW_BUCKET_COUNT; ++bucket) {
+        for (int i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
+            const auto id = vvNew[bucket][i];
```
f108193c6381b4d It looks like the loops can be accelerated with a faster iterator comparison and by avoiding integer conversions. 

<details><summary>suggestion</summary><p>

```diff
diff --git a/src/addrman.cpp b/src/addrman.cpp
index f2d54c0adb..41375e8320 100644
--- a/src/addrman.cpp
+++ b/src/addrman.cpp
@@ -703,8 +703,8 @@ std::vector<bool> CAddrMan::DecodeAsmap(fs::path path)
 
 void CAddrMan::ResetI2PPorts()
 {
-    for (int bucket = 0; bucket < ADDRMAN_NEW_BUCKET_COUNT; ++bucket) {
-        for (int i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
+    for (size_t bucket = 0; bucket != ADDRMAN_NEW_BUCKET_COUNT; ++bucket) {
+        for (size_t i = 0; i != ADDRMAN_BUCKET_SIZE; ++i) {
             const auto id = vvNew[bucket][i];
             if (id == -1) {
                 continue;
@@ -728,7 +728,7 @@ void CAddrMan::ResetI2PPorts()
             // re-evaluate that decision, but even if we could, CAddrInfo::GetNewBucket() does not
             // use CAddrInfo::GetKey() so it would end up in the same bucket as before the port
             // change.
-            const auto i_target = addr_info_newport.GetBucketPosition(nKey, true, bucket);
+            const size_t i_target = addr_info_newport.GetBucketPosition(nKey, true, bucket);
 
             if (i_target == i) { // No need to re-position.
                 addr_info = addr_info_newport;
@@ -743,8 +743,8 @@ void CAddrMan::ResetI2PPorts()
         }
     }
 
-    for (int bucket = 0; bucket < ADDRMAN_TRIED_BUCKET_COUNT; ++bucket) {
-        for (int i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
+    for (size_t bucket = 0; bucket != ADDRMAN_TRIED_BUCKET_COUNT; ++bucket) {
+        for (size_t i = 0; i != ADDRMAN_BUCKET_SIZE; ++i) {
             const auto id = vvTried[bucket][i];
             if (id == -1) {
                 continue;
@@ -763,8 +763,8 @@ void CAddrMan::ResetI2PPorts()
             // bucket and a position within that bucket. So a re-bucketing may be necessary.
             addr_info_newport.port = I2P_SAM31_PORT;
 
-            const auto bucket_target = addr_info_newport.GetTriedBucket(nKey, m_asmap);
-            const auto i_target = addr_info_newport.GetBucketPosition(nKey, false, bucket_target);
+            const size_t bucket_target = addr_info_newport.GetTriedBucket(nKey, m_asmap);
+            const size_t i_target = addr_info_newport.GetBucketPosition(nKey, false, bucket_target);
 
             if (bucket_target == bucket && i_target == i) { // No need to re-position.
                 addr_info = addr_info_newport;
```
</p></details>

jonatack replied:
(This isn't a hotspot but is something that could also be considered in places that are.)

laanwj replied:
Will leave it up to @vasild but it seems to me that integer and iterator conversions seem an unlikely source of performance problems.

> (This isn't a hotspot but is something that could also be considered in places that are.)

Sure, it might have impact on tight loops that loop over large memory areas.

vasild replied:
1.
`ADDRMAN_NEW_BUCKET_COUNT` is of type `int`.

```cpp
for (int bucket = 0; bucket < ADDRMAN_NEW_BUCKET_COUNT; ++bucket) {
```

how is changing `bucket` to `size_t` going to make it faster? (otoh, it could produce some warning about comparing `size_t` to `int`)

2.
Current code:
```cpp
int GetFoo();
auto x = GetFoo();
```
So `x` will be defined as type `int`. How will explicitly defining it as `size_t` make it faster? (otoh, assigning from `int` to `size_t` could cause some compiler to warn about it)

I tried this:

```cpp
#define N (1 << 10)

void foo()
{
    int array[N];
    for (int i = 0; i < N; ++i) {
        printf("%d\n", array[i]);
    }
}
```

Changing `i` from `int` to `size_t` produces **identical** assembler output:

```asm
(gdb) disas foo
Dump of assembler code for function _Z3foov:
   0x0000000000202220 <+0>:	push   %rbp
   0x0000000000202221 <+1>:	mov    %rsp,%rbp
   0x0000000000202224 <+4>:	push   %rbx
   0x0000000000202225 <+5>:	sub    $0x1008,%rsp
   0x000000000020222c <+12>:	xor    %ebx,%ebx
   0x000000000020222e <+14>:	xchg   %ax,%ax
   0x0000000000202230 <+16>:	mov    -0x1010(%rbp,%rbx,4),%esi
   0x0000000000202237 <+23>:	mov    $0x200cc7,%edi
   0x000000000020223c <+28>:	xor    %eax,%eax
   0x000000000020223e <+30>:	call   0x202670 <printf@plt>
   0x0000000000202243 <+35>:	add    $0x1,%rbx
   0x0000000000202247 <+39>:	cmp    $0x400,%rbx
   0x000000000020224e <+46>:	jne    0x202230 <_Z3foov+16>
   0x0000000000202250 <+48>:	add    $0x1008,%rsp
   0x0000000000202257 <+55>:	pop    %rbx
   0x0000000000202258 <+56>:	pop    %rbp
   0x0000000000202259 <+57>:	ret    
End of assembler dump.
```

vasild commented:
`0b0ee030ea...4101ec9d2e`: rebase due to conflicts, clarify `-port=` help wrt I2P.

jonatack commented on src/init.cpp:
```diff
@@ -442,7 +442,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-peerblockfilters", strprintf("Serve compact block filters to peers per BIP 157 (default: %u)", DEFAULT_PEERBLOCKFILTERS), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-permitbaremultisig", strprintf("Relay non-P2SH multisig (default: %u)", DEFAULT_PERMIT_BAREMULTISIG), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-port=<port>", strprintf("Listen for connections on <port>. Nodes not using the default ports (default: %u, testnet: %u, signet: %u, regtest: %u) are unlikely to get incoming connections.", defaultChainParams->GetDefaultPort(), testnetChainParams->GetDefaultPort(), signetChainParams->GetDefaultPort(), regtestChainParams->GetDefaultPort()), ArgsManager::ALLOW_ANY | ArgsManager: ...
+    argsman.AddArg("-port=<port>", strprintf("Listen for connections on <port>. Nodes not using the default ports (default: %u, testnet: %u, signet: %u, regtest: %u) are unlikely to get incoming connections. Not relevant for I2P (see doc/i2p.md).", defaultChainParams->GetDefaultPort(), testnetChainParams->GetDefaultPort(), signetChainParams->GetDefaultPort(), regtestChainParams->GetDefaultPort()) ...
```
```
$ ./src/bitcoind -h | grep -A4 "\-port="
  -port=<port>
       Listen for connections on <port>. Nodes not using the default ports
       (default: 8333, testnet: 18333, signet: 38333, regtest: 18444)
       are unlikely to get incoming connections. Not relevant for I2P
       (see doc/i2p.md).
```
maybe s/I2P/I2P connections/ and s/Not/This option is not/
```suggestion
    argsman.AddArg("-port=<port>", strprintf("Listen for connections on <port>. Nodes not using the default ports (default: %u, testnet: %u, signet: %u, regtest: %u) are unlikely to get incoming connections. This option is not relevant for I2P connections (see doc/i2p.md).", defaultChainParams->GetDefaultPort(), testnetChainParams->GetDefaultPort(), signetChainParams->GetDefaultPort(), regtestChainParams->GetDefaultPort()), ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
```

jonatack reviewed (commented):
re-ACK 4101ec9d2e05a35c35f587a28f1feee6cebcc61b per `git range-diff efff9c3 0b0ee03 4101ec9`, built with DDEBUG_ADDRMAN, did fairly extensive testing on mainnet both with and without a peers.dat / -dnsseeds=0 to test boostrapping.

All the I2P peers returned by `getaddednodeinfo` have port 0.

However, when bootstrapping with no peers.dat and some addnode peers, all but one of the peers returned by `getnodeaddresses 0 i2p` have port 8333, presumably due to the first addnode peer connected (with port 0) relaying addresses with that port number. (No ports were specified in the addnode config.)

jonatack commented:
Back from meeting; shut down the node, deleted the new peers.dat, recopied my backed-up peers.dat (`cp ~/.bitcoin/peers-bak.dat ~/.bitcoin/peers.dat`) and restarted.  Unlike bootstrapping with no peers.dat as reported just above in https://github.com/bitcoin/bitcoin/pull/22112#pullrequestreview-703116399, all of the I2P addresses in `getnodeaddresses 0 i2p` have port 0 (and as above, `getaddednodeinfo` as well).

```
2021-07-09T15:39:34Z Loaded 68521 addresses from peers.dat  2688ms
2021-07-09T15:39:34Z Loaded 2 addresses from "anchors.dat"
```
```
$ ./src/bitcoin-cli -rpcwait -addrinfo
{
  "addresses_known": {
    "ipv4": 50863,
    "ipv6": 7727,
    "torv2": 0,
    "torv3": 5557,
    "i2p": 16,
    "total": 64163
  }
}
```

jonatack commented:
Repeated the bootstrapping steps in https://github.com/bitcoin/bitcoin/pull/22112#pullrequestreview-703116399 twice more, with and without `-dnsseed=0`, same result.

jonatack commented:
Repeated the bootstrapping steps in https://github.com/bitcoin/bitcoin/pull/22112#pullrequestreview-703116399 but with no `addnode` peers and with `-dnsseed=0`.

```
2021-07-10T08:44:05Z Adding fixed seeds as -dnsseed=0, -addnode is not provided and all -seednode(s) attempted
```

jonatack commented:
Repeated the bootstrapping steps in https://github.com/bitcoin/bitcoin/pull/22112#pullrequestreview-703116399 but with no `addnode` peers and with `-dnsseed=0`.

```
2021-07-10T08:44:03Z Missing or invalid file ~/.bitcoin/peers.dat
2021-07-10T08:44:04Z DNS seeding disabled
2021-07-10T08:44:05Z Adding fixed seeds as -dnsseed=0, -addnode is not provided and all -seednode(s) attempted
```
```
        ipv4    ipv6   onion     i2p   total   block
in         0       0       0       2       2
out        9       0       1       0      10       2
total      9       0       1       2      12
```
```
$ ./src/bitcoin-cli -rpcwait -addrinfo
{
  "addresses_known": {
    "ipv4": 6281,
    "ipv6": 1076,
    "torv2": 0,
    "torv3": 58,
    "i2p": 5,
    "total": 7420
  }
}
$ ./src/bitcoin-cli getaddednodeinfo
[
]
```
All addresses returned by `getnodeaddresses 0 i2p` have port 0.

jonatack commented:
After a few minutes, one new `getnodeaddresses 0 i2p` entry from an inbound peer had port 8333. Restarted the node and that address now has port 0. However, still seeing the double connection issue (all report port 0); as is the case on master, it resolves within a minute or two.

![Screenshot from 2021-07-10 11-05-40](https://user-images.githubusercontent.com/2415484/125158129-715c9380-e15e-11eb-848e-accefe73e6f2.png)

vasild commented:
It is expected to have I2P addresses with port 8333 appearing in addrman - if some of our peers relays such one to us. After a restart the port will be changed to 0. It is important to remove this behavior (last commit from this PR) before real nodes appear that advertise themselves with port 8333 and only accept connections to that port (must be using SAM>=3.2 and deliberately close the connection if TO_PORT is not 8333).

It is also expected that a node running `master` can connect to a node running this PR two times as the fix is when initiating a connection (`master` does not have it). Here is how it looks like:

```sh
master$ bitcoin-cli addnode pr.b32.i2p:8333 onetry
master$ bitcoin-cli addnode pr.b32.i2p:0 onetry
master$ bitcoin-cli getpeerinfo |jq -r 'map(select(.addr |contains("pr.b32.i2p"))) |map({id: .id, addr: .addr, addrbind: .addrbind, inbound: .inbound})'
[
  {
    "id": 0,
    "addr": "pr.b32.i2p:8333",
    "addrbind": "master.b32.i2p:8333",
    "inbound": false
  },
  {
    "id": 1,
    "addr": "pr.b32.i2p:0",
    "addrbind": "master.b32.i2p:8333",
    "inbound": false
  }
]

pr$ bitcoin-cli getpeerinfo |jq -r 'map(select(.addr |contains("master.b32.i2p"))) |map({id: .id, addr: .addr, addrbind: .addrbind, inbound: .inbound})'
[
  {
    "id": 23708,
    "addr": "master.b32.i2p:0",
    "addrbind": "pr.b32.i2p:0",
    "inbound": true
  },
  {
    "id": 23717,
    "addr": "master.b32.i2p:0",
    "addrbind": "pr.b32.i2p:0",
    "inbound": true
  }
]

```

jonatack commented:
Yes, I came to the same conclusions.

FWIW tested the previous push and this one for 3 days each.

laanwj commented:
> It is important to remove this behavior (last commit from this PR) before real nodes appear that advertise themselves with port 8333 and only accept connections to that port (must be using SAM>=3.2 and deliberately close the connection if TO_PORT is not 8333).

That's pretty easy to forget. Let's open an issue for this once this is merged.

vasild commented:
> That's pretty easy to forget. Let's open an issue for this once this is merged.

I will take care about it. Maybe even a draft-PR that lingers for e.g. 6 months.

jonatack commented:
Yeah, I'd remind you ;)

Might be good to have this in rc1 to sanity test it with some other nodes running it.

laanwj commented:
Code review ACK 4101ec9d2e05a35c35f587a28f1feee6cebcc61b
Have also loosely tested it by running different versions of this on a node for the last weeks, during which I have noticed no problem with I2P connectivity.
